### ISSUE TITLE
Add support for a `testing` configuration option in cardano-node

### ISSUE THREAD

**Issue by nfrisby** [input-output-hk/ouroboros-network#2938]
The following code _statically_ determines which eras, mini protocol versions, ledger queries, and serialization schemes a deployment of `cardano-node` will use. Furthermore, there is some code in `cardano-node` that does the same for other versioned features (eg the "max supported `ProtVer`" that announced via an otherwise-unused field in the header of each minted block).

This Issue is to make these configurations partially dynamic. In particular, we intend to be able to add support in the Network Layer and Consensus Layer for a new miniprotocol version and/or a new ledger era etc but to do so _without necessarily_ enabling that support in a defaultly configured `cardano-node` instance. Instead, a new `cardano-node` configuration option, such as `EnableTesting` or `EnableExperimental` etc would be enabled by the eg Benchmarking and QA Teams when vetting the _next_ release.

(There is an analogous `NodeToClient` that also needs the same treatment, but I've omitted it from all of these snippets.)

https://github.com/input-output-hk/ouroboros-network/blob/e142c7153b9ff0505a84ec68124f5bd80449eebd/ouroboros-consensus/src/Ouroboros/Consensus/Node/NetworkProtocolVersion.hs#L38-L41

https://github.com/input-output-hk/ouroboros-network/blob/a9090675a7c63ae6369288faf91644fcb2c993fc/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L315-L323

https://github.com/input-output-hk/ouroboros-network/blob/1c4b3c7009abb9fd9390f7f61483f990c985cdb4/ouroboros-network/src/Ouroboros/Network/NodeToNode/Version.hs#L27-L51

https://github.com/input-output-hk/ouroboros-network/blob/a9090675a7c63ae6369288faf91644fcb2c993fc/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L216-L240

**Comment by nfrisby** [comment]
@dcoutts I wrote the above based on my understanding of our phone call last week.

Q1 Did I miss anything?

Q2 This seems like a reasonable compromise: we're add some flexibility to our versioning configuration so that we can continue to use a simple linear version management scheme -- both in the node-to-node negotation itself and also in our use of git branches etc when developing and/or cutting a release. Is that an accurate assessment of the underlying motivation?

**Comment by nfrisby** [comment]
@dcoutts Pointed me at https://github.com/input-output-hk/cardano-node/blob/master/cardano-node/src/Cardano/Node/Configuration/POM.hs#L115

And gave `TestMaryHardForkAtEpoch` as the kind of existing option name to emulate.

**Comment by nfrisby** [comment]
I've audited the code. `supportedNodeToNodeVersions` and the NTC analog only occur here:

https://github.com/input-output-hk/ouroboros-network/blob/2c63c74f1134ec2fe1a0a9dec6d84645b46ba3bf/ouroboros-consensus/src/Ouroboros/Consensus/Node.hs#L657-L660

Those fields are part of the "low-level" arguments to the node kernel.

https://github.com/input-output-hk/ouroboros-network/blob/2c63c74f1134ec2fe1a0a9dec6d84645b46ba3bf/ouroboros-consensus/src/Ouroboros/Consensus/Node.hs#L192-L196

So the `cardano-node` downstream change can provide those arguments a as restrictions of the "full" maps provided by the existing type classes instances.

In summary: `ouroboros-consensus` doesn't actually need to change here, besides promoting those `llrn` fields to be part of the standard interface.

The other half of this change (the per-era `ProtVer` header fields) are already specified in `cardano-node`.
